### DISCUSSION TITLE
UnknownAction type in listenerMiddleware when using matcher

### DISCUSSION THREAD

**Discussion by medihack** [reduxjs/redux-toolkit#4018] [Q&A] [answered]
As I update to RTK 2 I also rewrite my middleware to use `createListenerMiddleware`. In the documentation there is following example:

```ts
listenerMiddleware.startListening({
  matcher: isAnyOf(action1, action2, action3),
  effect: (action, listenerApi) => {
    const user = selectUserDetails(listenerApi.getState())

    const { specialData } = action.meta

    analyticsApi.trackUsage(action.type, user, specialData)
  },
})
```

When I do this and use `matcher` as in the example, then the `action` is typed as `UnknownAction` in the `effect` block. This is in contrast to when using `actionCreator`, where it is correctly typed with `payload` and `meta`. Is this the correct behavior? I would have expected some union type of all the actions given to the `matcher` and not an unknown action.

EDIT:

I also tried it with the Typescript example [given in the documentation](https://redux-toolkit.js.org/api/createListenerMiddleware#typescript-usage), but the action is still of type `UnknownAction` inside the `effect` block.

```ts
import { createListenerMiddleware, addListener, isAnyOf } from "@reduxjs/toolkit"
import type { TypedStartListening, TypedAddListener } from "@reduxjs/toolkit"
import { refreshMenu, setSelectedItem } from "./designerSlice"
import type { RootState, AppDispatch } from "./store"

export const listenerMiddleware = createListenerMiddleware()

export type AppStartListening = TypedStartListening<RootState, AppDispatch>

export const startAppListening = listenerMiddleware.startListening as AppStartListening

export const addAppListener = addListener as TypedAddListener<RootState, AppDispatch>

startAppListening({
  matcher: isAnyOf(refreshMenu, setSelectedItem),
  effect: (action, listenerApi) => {
    const { specialData } = action.meta
  },
})
```

So it seems RTK won't help me there and I have to cast the action meta myself. Is this correct?

**Comment by EskiMojo14** [comment, marked as answer]
we're not really sure why, but for some reason the inference order broke in 2.0.

You can force the correct order by defining the matcher as a separate variable first:
```ts
const isSpecialAction = isAnyOf(refreshMenu, setSelectedItem);
startAppListening({
  matcher: isSpecialAction,
  effect: (action, listenerApi) => {
    const { specialData } = action.meta
  },
})
```

**Reply by medihack** [reply]
That works great. The TypeScript gods are sometimes harsh and erratic for no apparent reason ;-) Thank you!

**Reply by jackdh** [reply]
I've noticed this error is still happening in version "^2.2.6", is there a bug report for this?
